storage-dynamodb/src/main/java/zipkin2/storage/dynamodb/DynamoDBSpanConsumer.java

```diff
@@ -0,0 +1,269 @@
+        spanPut.addItemEntry(ANNOTATIONS, new AttributeValue().withSS(
+            span.annotations().stream().map(Annotation::value).collect(Collectors.toList())
+        ));
+      }
+
+      if (span.timestamp() != null) {
+        spanPut.addItemEntry(SPAN_TIMESTAMP,
+            new AttributeValue().withN(String.valueOf(span.timestampAsLong())));
+      }
+
+      if (span.duration() != null) {
+        spanPut.addItemEntry(SPAN_DURATION,
+            new AttributeValue().withN(String.valueOf(span.durationAsLong())));
+      }
+
+      spanPut.addItemEntry(TTL_COLUMN, ttl);
+
+      result.add(new WriteRequest(spanPut));
+    }
+    return result;
+  }
+
+  private String timestampId(long timestamp) {
+    return BigInteger.valueOf(timestamp).shiftLeft(Long.SIZE).add(new BigInteger(Long.SIZE, random))
+        .toString();
+  }
+
+  private List<UpdateItemRequest> createUpsertsForServiceSpanNames(List<Span> spans,
+      AttributeValue ttl) {
+    Set<Pair> pairs = new HashSet<>();
+
+    for (Span span : spans) {
+      String spanName = StringUtils.isNullOrEmpty(span.name()) ? UNKNOWN : span.name();
+      pairs.add(new Pair(
+          StringUtils.isNullOrEmpty(span.localServiceName()) ? UNKNOWN : span.localServiceName(),
+          spanName));
+
+      if (!StringUtils.isNullOrEmpty(span.localServiceName())) {
+        pairs.add(new Pair(span.localServiceName(), WILDCARD_FOR_INVERTED_INDEX_LOOKUP));
+      }
+
+      if (!StringUtils.isNullOrEmpty(span.remoteServiceName())) {
+        pairs.add(new Pair(span.remoteServiceName(), spanName));
+        pairs.add(new Pair(span.remoteServiceName(), WILDCARD_FOR_INVERTED_INDEX_LOOKUP));
+      }
+    }
+
+    return pairs.stream()
+        .map(p -> p.asUpdateItemRequest(SERVICE_SPAN_ENTITY_TYPE, ttl))
+        .collect(Collectors.toList());
+  }
+
+  private List<UpdateItemRequest> createUpsertsForAutocompleteTags(List<Span> spans,
+      AttributeValue ttl) {
+    Set<Pair> pairs = new HashSet<>();
+
+    for (Span span : spans) {
+      for (Map.Entry<String, String> tag : span.tags().entrySet()) {
+        if (autocompleteKeys.contains(tag.getKey())) {
+          pairs.add(new Pair(tag.getKey(), tag.getValue()));
+          pairs.add(new Pair(tag.getKey(), WILDCARD_FOR_INVERTED_INDEX_LOOKUP));
+        }
+      }
+    }
+
+    return pairs.stream()
+        .map(p -> p.asUpdateItemRequest(AUTOCOMPLETE_TAG_ENTITY_TYPE, ttl))
+        .collect(Collectors.toList());
+  }
+
+  private static class Pair {
+    private String key;
+    private String value;
+
+    Pair(String key, String value) {
+      if (key == null) throw new IllegalArgumentException("Key cannot be null");
+      if (value == null) throw new IllegalArgumentException("Value cannot be null");
+      this.key = key;
+      this.value = value;
+    }
+
+    UpdateItemRequest asUpdateItemRequest(String type, AttributeValue ttl) {
+      return new UpdateItemRequest()
+          .withKey(new HashMap<String, AttributeValue>() {{
```

Comment:
`addKeyEntry` here can be used to avoid an anonymous `HashMap`

Reply:
not done I think